lib/Dialect/TTIR/Transforms/HoistCPUOps.cpp

```diff
@@ -0,0 +1,256 @@
+  const auto ranks = getTensorRanks(opToHoist);
+
+  const auto functionName = generateHoistedFuncName(opToHoist);
+
+  auto localFunc = llvm::dyn_cast_or_null<func::FuncOp>(
+      sourceModule.lookupSymbol(functionName));
+
+  // if we have not already emitted an equivalent function call, perform this
+  // hoist
+  if (localFunc == nullptr) {
+
+    mlir::MLIRContext *context = sourceModule.getContext();
+
+    // Gather operand and result types
+    llvm::SmallVector<mlir::Type> operandTypes, resultTypes;
+    for (auto operand : opToHoist->getOperands()) {
+      operandTypes.push_back(operand.getType());
+    }
+    for (auto result : opToHoist->getResultTypes()) {
+      resultTypes.push_back(result);
+    }
+
+    // Create the function signature
+    mlir::FunctionType funcType =
+        mlir::FunctionType::get(context, operandTypes, resultTypes);
+
+    // Create the function in the target module
+    auto hoistedFunc =
+        func::FuncOp::create(opToHoist->getLoc(), functionName, funcType);
+    targetModule.push_back(hoistedFunc);
+
+    // Add a basic block to the function
+    mlir::Block *block = hoistedFunc.addEntryBlock();
+    mlir::OpBuilder builder(block, block->end());
+
+    // Map operands to block arguments and clone the operation
+    llvm::SmallVector<mlir::Value> newOperands;
+    for (auto operand : llvm::enumerate(opToHoist->getOperands())) {
+      newOperands.push_back(block->getArgument(operand.index()));
+    }
+
+    mlir::IRMapping mapping;
+    for (auto operand : llvm::zip(opToHoist->getOperands(), newOperands)) {
+      mapping.map(std::get<0>(operand), std::get<1>(operand));
+    }
+
+    mlir::Operation *clonedOp = builder.clone(*opToHoist, mapping);
+
+    // Add a return operation to the function
+    builder.create<mlir::func::ReturnOp>(opToHoist->getLoc(),
+                                         clonedOp->getResults());
+
+    // Declare the function prototype in the source module
+    localFunc =
+        func::FuncOp::create(opToHoist->getLoc(), functionName, funcType);
+    localFunc.setPrivate(); // Mark as external (no body)
+    sourceModule.push_back(localFunc);
+
+    hoistedFunc->setAttr("arg_ranks", builder.getI64ArrayAttr(ranks));
+  }
+
+  // Replace the original operation with a call to the hoisted function
+  mlir::OpBuilder opBuilder(opToHoist);
+  auto callOp = opBuilder.create<mlir::func::CallOp>(
+      opToHoist->getLoc(), localFunc, opToHoist->getOperands());
+
+  // Replace all results of the original operation with the call results
+  for (auto result : llvm::zip(opToHoist->getResults(), callOp.getResults())) {
+    std::get<0>(result).replaceAllUsesWith(std::get<1>(result));
+  }
+
+  // Erase the original operation
+  opToHoist->erase();
+}
+
+// an analysis class which relies on manual input locations to mark ops to be
+// hoisted, useful for testing
+class TTIRHoistAnalyzeManual {
+public:
+  using HoistOpSet = llvm::SmallVector<llvm::SmallSet<mlir::Operation *, 4>>;
+
```

Comment:
Why not just use SmallVector for this. You are only interested into which ops you want to hoist.
Here you return list of sets where each set is of size 1

Reply:
Yeah, so the long-term vision is we might want to hoist sets of ops together, so this is the long-term signature we want.  That introduces a lot of complexity, and I'd rather get simple case working end-to-end first, so for now I'm only implementing single op hoisting, but multi-op hoisting is high on my TO-DO list.
Tracking as this issue: https://github.com/tenstorrent/tt-mlir/issues/1646

I could see argument we should simplify this to just be vector until then I guess, don't feel super strongly either way

Reply:
I see